### PR TITLE
Fixes #3520 - respect special chars width

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

File: lib/hammer_cli/output/adapter/table.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 require 'table_print'
 require File.join(File.dirname(__FILE__), 'wrapper_formatter')
+require File.join(File.dirname(__FILE__), '../../../table_print/formatter')
```
Comment (shlomizadok) [inline lib/hammer_cli/output/adapter/table.rb]: I honestly hate adding this `../../../` - any other way?

**Comment by ares** [comment]
\<unwanted tip> It would be also great to try fix it in upstream - https://github.com/arches/table_print/issues/2 \</unwanted tip>

File: lib/hammer_cli/output/adapter/table.rb (hunk 1 above)
Comment (ares) [inline lib/hammer_cli/output/adapter/table.rb]: doesn't just `'table_print/formatter'` work? if the gem is available I'd expect it to work

**Comment by shlomizadok** [comment]
This monkey patch is breaking tests.
@mbacovsky opened a patch on the gem at https://github.com/arches/table_print/pull/45
I think the only change I made was the gsub `/[\u0080-\u00ff]/`

Anyhow, we should try and have this on table_print and not here.

File: lib/hammer_cli/output/adapter/table.rb (hunk 1 above)
Comment (tstrachota) [inline lib/hammer_cli/output/adapter/table.rb]: I did some testing and the file `formatter.rb` is actually not being found at all. I'd move it from `lib` somewhere to 'lib/hammer_cli' and then just do
```ruby
require 'hammer_cli/...'
```

**Comment by tstrachota** [comment]
@shlomizadok I discovered a few things while testing this PR:

The tests are failing because they don't find the file with monkey patch. When I was testing it manually it was calculating length for chinese characters ok, but it failed for czech letters like "žščř". I found out that this is primarily because of false assumption in the original table_print code. The author expects every multibyte character to be rendered in two-columns, which is not correct (https://github.com/arches/table_print/blob/master/lib/table_print/formatter.rb#L44).

So I wrote couple of tests and dug deeper into the topic. I realized it's much more complex than one would expect :) I ended up with horrible Pascal-like algorithm that calculates correct length and is able to truncate strings containing both colours and two-column characters. You can find it all here:
https://github.com/shlomizadok/hammer-cli/pull/1

Can you please review it? I'll be glad for any suggestions for making the algorithm cleaner and more readable. I wasn't able to improve it any further at the point I was writing it:)

**Comment by domcleal** [comment]
> When I was testing it manually it was calculating length for chinese characters ok, but it failed for czech letters like "žščř". I found out that this is primarily because of false assumption in the original table_print code. The author expects every multibyte character to be rendered in two-columns, which is not correct

The [unicode-display_width gem](https://github.com/janlelis/unicode-display_width) provides a standard way of determining the display length, I'd strongly recommend using this instead of reimplementing it:

    irb(main):001:0> Unicode::DisplayWidth.of("žščř")
    => 4

**Comment by tstrachota** [comment]
Nice, it's surely better to use existing solution.

**Comment by tstrachota** [comment]
Updated. Unfortunately it can't cope with colours so some custom code still remains.

**Comment by tstrachota** [comment]
Merging that disqualifies me from doing the review :) May I ask somebody else? @mbacovsky or @ares ?

**Comment by ares** [comment]
It seems some characters still breaks it :-(

I added second name column in hammer-cli-foreman to see if it's aligned properly...
hammer

**Comment by tstrachota** [comment]
@ares you're right, that was introduced after I switched to `unicode-display_width`. Before that commit it works fine for czech letters. I have to add such values to the tests.

**Comment by tstrachota** [comment]
@ares updated in https://github.com/shlomizadok/hammer-cli/pull/2

**Comment by ares** [comment]
I checkout your branch and it works just fine, :+1: ACK once Shlomi merges the PR, I guess some packaging will be required once this gets in.

**Comment by mmoll** [comment]
@ares only the unicode-display_width gem needs to get packaged and hammer updated, or am I missing something? :)

**Comment by ares** [comment]
That's what I meant, I just wanted to kindly remind authors to open packaging PRs :-)

**Comment by mmoll** [comment]
what's the status here?

**Comment by tstrachota** [comment]
I found out we already have rpm package for unicode-display_width:
https://github.com/theforeman/foreman-packaging/tree/rpm/develop/rubygem-unicode-display_width

We only need debs.

**Comment by mmoll** [comment]
@tstrachota the DEB PR is there and marked ready to merge. :)

**Comment by tstrachota** [comment]
@mmoll nice! I completely missed that. Thank you!

**Comment by tstrachota** [comment]
[tests]

**Comment by tstrachota** [comment]
The last thing blocking the merge is tests. @shlomizadok could you please rebase? It seems that jenkins fails to apply the patches.

**Comment by shlomizadok** [comment]
@tstrachota rebased :)

**Comment by mmoll** [comment]
💚

**Comment by tstrachota** [comment]
Merged, thanks @shlomizadok and @mmoll !

**Pull request merged**